File: programming-models/webflux-fn/src/main/java/com/example/demo/DemoApplication.java
package com.example.demo;

import io.r2dbc.spi.Row;
import io.r2dbc.spi.RowMetadata;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.annotation.Id;
import org.springframework.data.r2dbc.core.R2dbcEntityTemplate;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;
import org.springframework.data.relational.core.query.Update;
import org.springframework.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.net.URI;
import java.time.LocalDateTime;
import java.util.function.BiFunction;

import static org.springframework.data.relational.core.query.Criteria.where;
import static org.springframework.data.relational.core.query.Query.query;
import static org.springframework.http.HttpMethod.*;
import static org.springframework.web.reactive.function.server.RequestPredicates.method;
import static org.springframework.web.reactive.function.server.RequestPredicates.path;
import static org.springframework.web.reactive.function.server.RouterFunctions.nest;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;
import static org.springframework.web.reactive.function.server.ServerResponse.*;

@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }

}

@Component
@RequiredArgsConstructor
@Slf4j
class SampleRunner implements ApplicationRunner {
    final PostRepository posts;

    @Override
    public void run(ApplicationArguments args) throws Exception {
        posts.findAll()
                .subscribe(
                        data -> log.info("initial post data: {}", data),
                        err -> log.error("err: {}", err)
                );
    }
}


@Configuration
class WebConfig {

    @Bean
    public RouterFunction<ServerResponse> routes(PostHandler postsHandler) {
        var collectionRoutes = route(method(GET), postsHandler::findAll)
                .andRoute(method(POST), postsHandler::create);
        var singleRoutes = route(method(GET), postsHandler::findById)
                .andRoute(method(PUT), postsHandler::update)
                .andRoute(method(DELETE), postsHandler::deleteById);

        return route()
                .path("posts",
                        () -> nest(path("{id}"), singleRoutes)
                                .andNest(path(""), collectionRoutes)
                )
                .build();
    }
}

@Component
@RequiredArgsConstructor
class PostHandler {

    private final PostRepository posts;

    public Mono<ServerResponse> findAll(ServerRequest req) {
        return ok().body(this.posts.findAll(), Post.class);
    }

    public Mono<ServerResponse> create(ServerRequest req) {
        return req.bodyToMono(Post.class)
                .flatMap(this.posts::create)
                .flatMap(postId -> created(URI.create("/posts/" + postId)).build());
    }

    public Mono<ServerResponse> findById(ServerRequest req) {
        return this.posts.findById(Long.valueOf(req.pathVariable("id")))
                .flatMap(post -> ok().body(Mono.just(post), Post.class))
                .switchIfEmpty(notFound().build());
    }

    public Mono<ServerResponse> update(ServerRequest req) {
        return req.bodyToMono(Post.class)
                .flatMap(p -> this.posts.update(Long.valueOf(req.pathVariable("id")), p))
                .flatMap(d -> {
                    if (d > 0) return noContent().build();
                    else return notFound().build();
                });
    }

    public Mono<ServerResponse> deleteById(ServerRequest req) {
        return this.posts.deleteById(Long.valueOf(req.pathVariable("id")))
                .flatMap(d -> {
                    if (d > 0) return noContent().build();
                    else return notFound().build();
                });
    }
}

@Component
@RequiredArgsConstructor
class PostRepository {

    private final DatabaseClient client;

    public static final BiFunction<Row, RowMetadata, Post> ROW_MAPPER = (row, meta) -> new Post(
            row.get("id", Long.class),
            row.get("title", String.class),
            row.get("content", String.class),
            row.get("created_at", LocalDateTime.class)
    );

    public Flux<Post> findAll() {
        String sql = """
            SELECT id, title, content, created_at
            FROM posts
        """;
        return client.sql(sql)
                .map(ROW_MAPPER)
                .all();
    }

    public Flux<Post> findByTitleLike(String title) {
        String sql = """
            SELECT id, title, content, created_at
            FROM posts
            WHERE title LIKE :title
        """;
        return client.sql(sql)
                .bind("title", "%" + title + "%")
                .map(ROW_MAPPER)
                .all();
    }

    public Mono<Post> findById(Long id) {
        String sql = """
            SELECT id, title, content, created_at
            FROM posts
            WHERE id = :id
        """;
        return client.sql(sql)
                .bind("id", id)
                .map(ROW_MAPPER)
                .one();
    }

    public Mono<Long> create(Post data) {
        String sql = """
            INSERT INTO posts (title, content)
            VALUES (:title, :content)
            RETURNING id
        """;
        return client.sql(sql)
                .bind("title", data.title())
                .bind("content", data.content())
                .map((row, meta) -> row.get("id", Long.class))
                .one();
    }

    public Mono<Long> update(Long id, Post data) {
        String sql = """
            UPDATE posts
            SET title = :title, content = :content
            WHERE id = :id
        """;
        return client.sql(sql)
                .bind("title", data.title())
                .bind("content", data.content())
                .bind("id", id)
                .fetch()
                .rowsUpdated();
    }

    public Mono<Long> deleteById(Long id) {
        String sql = """
            DELETE FROM posts
            WHERE id = :id
        """;
        return client.sql(sql)
                .bind("id", id)
                .fetch()
                .rowsUpdated();
    }
}

@Table(value = "posts")
record Post(
        @Id
        @Column("id")
        Long id,

        @Column("title")
        String title,

        @Column("content")
        String content,

        @Column("created_at")
        LocalDateTime createdAt
) {
}

File: programming-models/webflux-ktco-fn/src/main/kotlin/com/example/demo/DemoApplication.kt
package com.example.demo

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.reactive.asFlow
import kotlinx.coroutines.reactor.awaitSingle
import kotlinx.coroutines.reactor.awaitSingleOrNull
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.context.event.ApplicationReadyEvent
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.event.EventListener
import org.springframework.data.annotation.Id
import org.springframework.data.r2dbc.core.R2dbcEntityTemplate
import org.springframework.data.r2dbc.core.delete
import org.springframework.data.r2dbc.core.insert
import org.springframework.data.r2dbc.core.select
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import org.springframework.data.relational.core.query.Criteria.where
import org.springframework.data.relational.core.query.Query
import org.springframework.data.relational.core.query.Update
import org.springframework.http.MediaType
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.*
import org.springframework.web.reactive.function.server.ServerResponse.*
import java.net.URI
import java.time.LocalDateTime

@SpringBootApplication
class DemoApplication

fun main(args: Array<String>) {
    runApplication<DemoApplication>(*args)
}

@Component
class DataInitializer(private val posts: PostRepository) {

    @EventListener(value = [ApplicationReadyEvent::class])
    suspend fun init() {
        println(" print initial data...")
        posts.findAll().collect { println(it) }
    }

}

@Configuration
class WebConfig {

    @Bean
    fun routes(postHandler: PostHandler) = coRouter {
        "/posts".nest {
            "{id}".nest {
                GET(accept(MediaType.APPLICATION_JSON), postHandler::get)
                PUT(contentType(APPLICATION_JSON), postHandler::update)
                DELETE("", postHandler::delete)
            }
            GET(accept(MediaType.APPLICATION_JSON), postHandler::all)
            POST(contentType(APPLICATION_JSON), postHandler::create)
        }
    }
}

@Component
class PostHandler(private val posts: PostRepository) {

    suspend fun all(req: ServerRequest): ServerResponse {
        return ok().bodyAndAwait(this.posts.findAll())
    }

    suspend fun create(req: ServerRequest): ServerResponse {
        val body = req.awaitBody<CreatePostRequest>()
        val createdPost = this.posts.create(body)
        return created(URI.create("/$createdPost")).buildAndAwait()
    }

    suspend fun get(req: ServerRequest): ServerResponse {
        println("path variable::${req.pathVariable("id")}")
        val foundPost = this.posts.findById(req.pathVariable("id").toLong())
        println("found post:$foundPost")
        return when {
            foundPost != null -> ok().bodyValueAndAwait(foundPost)
            else -> notFound().buildAndAwait()
        }
    }

    suspend fun update(req: ServerRequest): ServerResponse {
        val id = req.pathVariable("id").toLong()
        val body = req.awaitBody<UpdatePostRequest>()
        val updated = this.posts.update(id, body)
        return when {
            updated > 0 -> noContent().buildAndAwait()
            else -> notFound().buildAndAwait()
        }
    }

    suspend fun delete(req: ServerRequest): ServerResponse {
        val deletedCount = this.posts.deleteById(req.pathVariable("id").toLong())
        println("$deletedCount posts deleted")
        return when {
            deletedCount > 0 -> noContent().buildAndAwait()
            else -> notFound().buildAndAwait()
        }
    }
}

@Component
class PostRepository(private val template: R2dbcEntityTemplate) {

    suspend fun count(): Long =
        template.select<Post>().count().awaitSingle()

    fun findAll(): Flow<Post> =
        template.select<Post>()
            .all()
            .asFlow()

    suspend fun findById(id: Long): Post? =
        template.selectOne(Query.query(where("id").`is`(id)), Post::class.java)
            .awaitSingleOrNull()

    suspend fun deleteById(id: Long): Long =
        template.delete(Query.query(where("id").`is`(id)), Post::class.java)
            .awaitSingle()

    suspend fun deleteAll(): Long =
        template.delete<Post>()
            .all()
            .awaitSingle()

    suspend fun create(post: CreatePostRequest): Long {
        val data = Post(title = post.title, content = post.content)
        return template.insert<Post>()
            .using(data)
            .map { it.id!! }
            .awaitSingle()
    }


    suspend fun update(id: Long, post: UpdatePostRequest): Long =
        template.update(
            Query.query(where("id").`is`(id)),
            Update.update("title", post.title)
                .set("content", post.content),
            Post::class.java
        ).awaitSingle()

    suspend fun init() {
        //client.execute().sql("CREATE TABLE IF NOT EXISTS posts (login varchar PRIMARY KEY, firstname varchar, lastname varchar);").await()
        val deletedCount = deleteAll()
        println(" $deletedCount posts deleted!")
        create(CreatePostRequest(title = "My first post title", content = "Content of my first post"))
        create(CreatePostRequest(title = "My second post title", content = "Content of my second post"))
    }
}

data class CreatePostRequest(val title: String, val content: String)

data class UpdatePostRequest(val title: String, val content: String)

@Table("posts")
data class Post(
    @Id val id: Long? = null,
    @Column("title") val title: String? = null,
    @Column("content") val content: String? = null,
    @Column("created_at") val createdAt: LocalDateTime? = null
)


